File: src/main/java/fyordo/lifeagragator/med/tag/TagController.java
package fyordo.lifeagragator.med.tag;

import fyordo.lifeagragator.med.base.reponse.ListResponse;
import fyordo.lifeagragator.med.tag.dto.TagDto;
import fyordo.lifeagragator.med.tag.request.TagCreateRequest;
import fyordo.lifeagragator.med.tag.request.TagFilter;
import fyordo.lifeagragator.med.tag.request.TagUpdateRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@Controller()
@RequiredArgsConstructor
@RequestMapping("/tag")
public class TagController {
    private final TagService tagService;
    @GetMapping("")
    public ResponseEntity<ListResponse<TagDto>> getAllTags(@RequestParam Map<String, String> filter){
        TagFilter tagFilter = new TagFilter(filter);
        List<TagDto> result = tagService.getTags(tagFilter).stream().map(TagDto::new).toList();
        return ResponseEntity.ok(new ListResponse<>(result));
    }

    @GetMapping("{id}")
    public ResponseEntity<TagDto> getTagById(@PathVariable Long id){
        return ResponseEntity.ok(new TagDto(tagService.getTagById(id)));
    }

    @PostMapping()
    public ResponseEntity<TagDto> createTag(@RequestBody TagCreateRequest request){
        return ResponseEntity.status(201).body(
                new TagDto(tagService.createTag(request))
        );
    }

    @PutMapping()
    public ResponseEntity<TagDto> updateTag(@RequestBody TagUpdateRequest request){
        return ResponseEntity.status(201).body(
                new TagDto(tagService.updateTag(request))
        );
    }

    @DeleteMapping("{id}")
    public ResponseEntity<?> deleteTag(@PathVariable Long id){
        tagService.deleteTagById(id);
        return ResponseEntity.status(204).build();
    }
}

File: src/main/kotlin/fyordo/lifeagragator/med/tag/TagController.kt
package fyordo.lifeagragator.med.tag

import fyordo.lifeagragator.med.tag.dto.TagCreateRequest
import fyordo.lifeagragator.med.tag.dto.TagDeleteMultipleRequest
import fyordo.lifeagragator.med.tag.dto.TagDto
import fyordo.lifeagragator.med.tag.dto.TagUpdateRequest
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/tag")
class TagController(
    private val tagService: TagService
) {
    @GetMapping
    fun findAll(): ResponseEntity<List<TagDto>> =
        ResponseEntity.ok(tagService.findAll().map { TagDto(it) })

    @GetMapping("/{id}")
    fun findById(@PathVariable id: Long): ResponseEntity<TagDto> =
        ResponseEntity.ok(TagDto(tagService.findById(id)))

    @PostMapping
    fun createTag(@RequestBody request: TagCreateRequest): ResponseEntity<TagDto> {
        return ResponseEntity.status(201).body(TagDto(tagService.createTag(request)))
    }

    @PutMapping
    fun updateTag(@RequestBody request: TagUpdateRequest): ResponseEntity<TagDto>{
        return ResponseEntity.status(200).body(TagDto(tagService.updateTag(request)))
    }

    @DeleteMapping("{id}")
    fun deleteTag(@PathVariable id: Long): ResponseEntity<Unit>{
        tagService.deleteTag(id)
        return ResponseEntity.noContent().build()
    }

    @PostMapping("delete-multiple")
    suspend fun deleteTags(@RequestBody request: TagDeleteMultipleRequest): ResponseEntity<Unit>{
        tagService.deleteTags(request.ids)
        return ResponseEntity.noContent().build()
    }
}